### PR TITLE
Redeem page

### LINKED ISSUE #104: Redeem Coupon Pages (Redeem Page 1 and 3)

### PR DESCRIPTION
**Pull request by sivayogasubramanian**
To be merged after #103.

Closes #104 

- [x] Create secondary donation frontend api
- [x] Use common components in charity selection page and complete implementation
- [x] Create a simple formik form for the amount for the amount page
- [x] Integrate buttons with api
- [x] Double check creation of secondary donation logic with team
- [x] Clean up code and abstract if needed

### CODE REVIEW

File: backend/app/views/secondary_donations/_secondary_donation.json.jbuilder [hunk 1]
```diff
@@ -1,6 +1,11 @@
...
 json.id secondary_donation.id
-json.couponId secondary_donation.coupon_id
 json.amount secondary_donation.amount
 json.campaignCharityId secondary_donation.campaign_charity_id
+
+if secondary_donation.coupon
+  json.urlToken secondary_donation.coupon.url_token
+else
+  json.urlToken nil
+end
```
Comment (zognin) [inline backend/app/views/secondary_donations/_secondary_donation.json.jbuilder:11]: why this change? does it affect anything?

File: frontend/components/redeem/CampaignCharitySelection.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  campaignCharityId: Nullable<number>;
+  setCampaignCharityId: (campaignCharityId: number) => void;
+  goToNextPage: () => void;
+};
+
+const CampaignCharitySelection = ({ coupon, campaignCharityId, setCampaignCharityId, goToNextPage }: Props) => {
+  return (
+    <Grid container sx={containerSx} component="main" justifyContent="center" padding={2}>
+      <Grid item md={12} lg={4}>
+        <Stack spacing={theme.spacing(2)} paddingBottom={4}>
```
Comment (zognin) [inline frontend/components/redeem/CampaignCharitySelection.tsx:22]: I realised the proper way is to use useTheme, and pass theme into SxProps objects through their function if needed. This is so we can swap themes easily like dark vs light theme through the `ThemeProvider` at `app.tsx`. Not changing the styles files yet, but for .tsx files let's use useTheme.

```suggestion
const CampaignCharitySelection = ({ coupon, campaignCharityId, setCampaignCharityId, goToNextPage }: Props) => {
  const theme = useTheme();
  return (
    <Grid container sx={containerSx} component="main" justifyContent="center" padding={2}>
      <Grid item md={12} lg={4}>
        <Stack spacing={theme.spacing(2)} paddingBottom={4}>
```

File: frontend/pages/redeem/[urlToken]/index.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import { Nullable } from '../../../types/utils';
+
+const INITIAL_REDEEM_PAGE = 1;
+const NUMBER_OF_REDEEM_PAGES = 2;
+
+const Redeem: NextPage = () => {
+  const router = useRouter();
+  const urlToken = router.query.urlToken && isString(router.query.urlToken) ? router.query.urlToken : null;
+  const { data: coupon, error } = useSWR<Nullable<CouponRedeemData>>([urlToken], (urlToken) =>
+    urlToken !== null ? api.coupons.getCoupon(urlToken).then((r) => r.payload) : null,
```
Comment (zognin) [inline frontend/pages/redeem/[urlToken]/index.tsx:23]: We should have `getUnredeemedCoupon` to only allow people to access unredeemed coupons. Maybe this can be done in another pr but it's a priority.

File: frontend/styles/redeem/indexStyles.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { SxProps } from '@mui/material';
+
+export const containerSx: SxProps = {
+  maxWidth: '1200px',
```
Comment (zognin) [inline frontend/styles/redeem/indexStyles.ts:4]: How was this picked?

**Review by zognin — COMMENTED** [review]
photo_2022-10-16 12 42 28


Please fix, the grid item should have a fixed height, not the grid container. Refer to `/components/CampaignList.tsx` line 19

Also I thought visit page was supposed to go to the charity's website not the givingsg campaign?

And that "Make a personal contribution" should go to the givingsg campaign because we want to test whether they'll come back

File: frontend/pages/redeem/[urlToken]/index.tsx (hunk 3 above)
Comment (sivayogasubramanian) [reply frontend/pages/redeem/[urlToken]/index.tsx:23]: Will be done in another PR. #172

File: frontend/styles/redeem/indexStyles.ts (hunk 4 above)
Comment (sivayogasubramanian) [reply frontend/styles/redeem/indexStyles.ts:4]: https://github.com/Giving-Coupons/giving-coupons/pull/66#discussion_r990742075

File: backend/app/views/secondary_donations/_secondary_donation.json.jbuilder (hunk 1 above)
Comment (sivayogasubramanian) [reply backend/app/views/secondary_donations/_secondary_donation.json.jbuilder:11]: Checked the frontend and nowhere do we use couponId. Created #173

**Comment by sivayogasubramanian** [comment]
Screenshot 2022-10-16 at 2 31 42 PM

Screenshot 2022-10-16 at 2 31 55 PM

**Comment by sivayogasubramanian** [comment]
Created #175 for the links. Can be done later since it is not urgent.

Fixed heights of cards in this PR.

The bar chart will be fixed at the component level in #103.

**Review by zognin — APPROVED** [review]
LGTM
